api/layer/multipart_upload.go

```diff
@@ -0,0 +1,566 @@
+		Delimiter      string
+		EncodingType   string
+		KeyMarker      string
+		MaxUploads     int
+		Prefix         string
+		UploadIDMarker string
+	}
+
+	ListPartsParams struct {
+		Info             *UploadInfoParams
+		MaxParts         int
+		PartNumberMarker int
+	}
+
+	ListPartsInfo struct {
+		Parts                []*Part
+		Owner                *owner.ID
+		NextPartNumberMarker int
+		IsTruncated          bool
+	}
+
+	ListMultipartUploadsInfo struct {
+		Prefixes           []string
+		Uploads            []*UploadInfo
+		IsTruncated        bool
+		NextKeyMarker      string
+		NextUploadIDMarker string
+	}
+	UploadInfo struct {
+		IsDir    bool
+		Key      string
+		UploadID string
+		Owner    *owner.ID
+		Created  time.Time
+	}
+)
+
+func (n *layer) UploadPart(ctx context.Context, p *UploadPartParams) (*data.ObjectInfo, error) {
+	if p.PartNumber != 0 {
+		if _, err := n.GetUploadInitInfo(ctx, p.Info); err != nil {
+			return nil, err
+		}
+	}
+
+	if p.Size > uploadMaxSize {
+		return nil, errors.GetAPIError(errors.ErrEntityTooLarge)
+	}
+
+	if p.Header == nil {
+		p.Header = make(map[string]string)
+	}
+
+	appendUploadHeaders(p.Header, p.Info.UploadID, p.Info.Key, p.PartNumber)
+
+	params := &PutObjectParams{
+		Bucket: p.Info.Bkt.Name,
+		Object: createUploadPartName(p.Info.UploadID, p.Info.Key, p.PartNumber),
+		Size:   p.Size,
+		Reader: p.Reader,
+		Header: p.Header,
+	}
+
+	return n.objectPut(ctx, p.Info.Bkt, params)
+}
+
+func (n *layer) UploadPartCopy(ctx context.Context, p *UploadCopyParams) (*data.ObjectInfo, error) {
+	if _, err := n.GetUploadInitInfo(ctx, p.Info); err != nil {
+		return nil, err
+	}
+
+	if p.Range != nil {
+		if p.Range.End-p.Range.Start > uploadMaxSize {
+			return nil, errors.GetAPIError(errors.ErrEntityTooLarge)
+		}
+		if p.Range.End > uint64(p.SrcObjInfo.Size) {
+			return nil, errors.GetAPIError(errors.ErrInvalidCopyPartRangeSource)
+		}
+	} else {
+		if p.SrcObjInfo.Size > uploadMaxSize {
+			return nil, errors.GetAPIError(errors.ErrEntityTooLarge)
+		}
+	}
+
+	metadata := make(map[string]string)
+	appendUploadHeaders(metadata, p.Info.UploadID, p.Info.Key, p.PartNumber)
+
+	c := &CopyObjectParams{
+		SrcObject: p.SrcObjInfo,
+		DstBucket: p.Info.Bkt.Name,
+		DstObject: createUploadPartName(p.Info.UploadID, p.Info.Key, p.PartNumber),
+		SrcSize:   p.SrcObjInfo.Size,
+		Header:    metadata,
+		Range:     p.Range,
+	}
+
+	return n.CopyObject(ctx, c)
+}
+
+func (n *layer) CompleteMultipartUpload(ctx context.Context, p *CompleteMultipartParams) (*data.ObjectInfo, error) {
+	var obj *data.ObjectInfo
+
+	for i := 1; i < len(p.Parts); i++ {
+		if p.Parts[i].PartNumber <= p.Parts[i-1].PartNumber {
+			return nil, errors.GetAPIError(errors.ErrInvalidPartOrder)
+		}
+	}
+
+	objects, err := n.getUploadParts(ctx, p.Info)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(objects) == 1 {
+		obj, err := n.headLastVersionIfNotDeleted(ctx, p.Info.Bkt, p.Info.Key)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
I think we should check err and if it's `NoSuchKey` log it and return `ErrInvalidPart` (or somehow change error to return). It's a little confusing when you try to invoke `complete-multipart-upload` and get `NoSuchKey`.
People can think they provide wrong key (at least I thought so) though they just tried to complete upload without any uploaded parts in neofs.
```
$ aws s3api complete-multipart-upload --endpoint http://127.0.0.1:8084  --bucket dkirillov2 --key mult1 --upload-id bd86067e-6d53-4967-8589-7340a79ea98b --multipart-upload file://mpustruct
An error occurred (NoSuchKey) when calling the CompleteMultipartUpload operation (reached max retries: 0): The specified key does not exist.
```

AWS returns `InvalidPart` error as I understand